### PR TITLE
Add land/lor/lxor for BigInt.

### PR DESCRIPTION
**Pull request by SyoujyoujiNaiki**
#1159

### CODE REVIEW

**Comment by SyoujyoujiNaiki** [comment]
cc @Lampese

**Comment by peter-jerry-ye** [comment]
Will check the pipeline issue.

However, I'm not quite sure about the implementation:

```bash
% node
Welcome to Node.js v23.4.0.
Type ".help" for more information.
> (-10n) & 11n
2n
> (-10n) | 11n
-1n
> (-10n) ^ 11n
-3n
```

**Comment by SyoujyoujiNaiki** [comment]
> Will check the pipeline issue.
> 
> However, I'm not quite sure about the implementation:
> 
> ```shell
> % node
> Welcome to Node.js v23.4.0.
> Type ".help" for more information.
> > (-10n) & 11n
> 2n
> > (-10n) | 11n
> -1n
> > (-10n) ^ 11n
> -3n
> ```

The implementation of NodeJS BigInt is different from the MoonBit NonJS implementation. I just apply the land/lor/lxor method on the uint array and the sign.

**Comment by peter-jerry-ye** [comment]
> The implementation of NodeJS BigInt is different from the MoonBit NonJS implementation. I just apply the land/lor/lxor method on the uint array and the sign.

I noticed that. But MoonBit needs consistency across backends, i.e. NonJS & JS backends.

**Comment by SyoujyoujiNaiki** [comment]
> > The implementation of NodeJS BigInt is different from the MoonBit NonJS implementation. I just apply the land/lor/lxor method on the uint array and the sign.
> 
> I noticed that. But MoonBit needs consistency across backends, i.e. NonJS & JS backends.

But then the land/lor/lxor would be not a direct and clear bit operation.

**Comment by peter-jerry-ye** [comment]
> But then the land/lor/lxor would be not a direct and clear bit operation.

It has never been: https://262.ecma-international.org/15.0/index.html#sec-bigintbitwiseop

File: builtin/bigint_nonjs.mbt [hunk 1]
```diff
@@ -876,3 +876,41 @@ pub fn BigInt::to_octets(self : BigInt, length? : Int) -> Bytes {
   }
   result
 }
+
+///|
+pub fn land(self : BigInt, other : BigInt) -> BigInt {
+  let new_limbs = FixedArray::make(self.limbs.length(), 0U)
+  for i = 0; i < self.len; i = i + 1 {
+    new_limbs[i] = self.limbs[i] & other.limbs[i]
```
Comment (inkydragon) [inline builtin/bigint_nonjs.mbt:884]: The implementation does not seem to take into account the length of the two inputs.
With
```
  // large input
  let a = BigInt::from_string("123456789012345678901234567890")
  inspect!(a & 0N, content="0")
```
We have: `RuntimeError: array element access out of bounds`

**Review by inkydragon — COMMENTED** [review]
Maybe we could learn something from quickjs (MIT).

`quickjs` uses a struct for representing a BigInt, moonbit doesn't have an `expn` field, and a similar implementation of `bf_logic_op` would be more compact. 
```c
/* +/-zero is represented with expn = BF_EXP_ZERO and len = 0,
   +/-infinity is represented with expn = BF_EXP_INF and len = 0,
   NaN is represented with expn = BF_EXP_NAN and len = 0 (sign is ignored)
 */
typedef struct {
    struct bf_context_t *ctx;
    int sign;
    slimb_t expn;
    limb_t len;
    limb_t *tab;
} bf_t;
```
https://github.com/quickjs-ng/quickjs/blob/be2db73000765717f231e32a4110cb72556d0bec/libbf.c#L2345-L2447

**Comment by peter-jerry-ye** [comment]
You might need something like this:

```moonbit
pub fn BigInt::land(self : BigInt, other : BigInt) -> BigInt {
  let mut x = self
  let mut y = other
  let mut result = 0N
  let mut shift = 0
  while not((x == 0 || x == -1) && (y == 0 || y == -1)) {
    let x0 = x.limbs[0] & 1
    let y0 = y.limbs[0] & 1
    if (x0 & y0) != 0 {
      result = result + (1N << shift)
    }
    x = (x - BigInt::from_int(x0.reinterpret_as_int())) / 2
    y = (y - BigInt::from_int(y0.reinterpret_as_int())) / 2
    shift += 1
  }
  if (x.limbs[0] & 1U & (y.limbs[0] & 1U)) != 0 {
    result = result - (1N << shift)
  }
  result
}
```

**Comment by SyoujyoujiNaiki** [comment]
> > The implementation of NodeJS BigInt is different from the MoonBit NonJS implementation. I just apply the land/lor/lxor method on the uint array and the sign.
> 
> I noticed that. But MoonBit needs consistency across backends, i.e. NonJS & JS backends.

Pushed the new implementation which results the same as what NodeJS output.

File: builtin/bigint_nonjs.mbt (hunk 1 above)
Comment (SyoujyoujiNaiki) [reply builtin/bigint_nonjs.mbt]: Pushed the new implementation that makes two BigInt's limbs when calculating same size long. Should fix this.

File: builtin/bigint_js.mbt [hunk 2]
```diff
@@ -212,3 +212,15 @@ extern "js" fn js_shl(self : BigInt, other : Int) -> BigInt =
...
+extern "js" fn js_land(self : BigInt, other : BigInt) -> BigInt =
+  #|(x, y) => x & y
+
+///|
+extern "js" fn js_lor(self : BigInt, other : BigInt) -> BigInt =
+  #|(x, y) => x | y
+
+///|
+extern "js" fn js_lxor(self : BigInt, other : BigInt) -> BigInt =
+  #|(x, y) => x ^ y
```
Comment (peter-jerry-ye) [inline builtin/bigint_js.mbt:226]:
```suggestion

///|
pub extern "js" fn land(self : BigInt, other : BigInt) -> BigInt =
  #|(x, y) => x & y

///|
pub extern "js" fn lor(self : BigInt, other : BigInt) -> BigInt =
  #|(x, y) => x | y

///|
pub extern "js" fn lxor(self : BigInt, other : BigInt) -> BigInt =
  #|(x, y) => x ^ y
```

**Review by peter-jerry-ye — CHANGES_REQUESTED** [review]
This should make the tests pass

File: builtin/bigint_nonjs.mbt [hunk 3]
```diff
@@ -925,3 +925,264 @@ pub fn BigInt::to_octets(self : BigInt, length? : Int) -> Bytes {
...
+      if x_limbs[i] != 0xFFFFFFFFU {
+        break
+      }
+    }
+    for i = 0; i < x_limbs.length(); i = i + 1 {
+      x_limbs[i] = x_limbs[i] ^ 0xFFFFFFFFU
+    }
+  }
+  
+  // return the result
```
Comment (peter-jerry-ye) [inline builtin/bigint_nonjs.mbt:999]:
```suggestion
  }

  // return the result
```

File: builtin/bigint_nonjs.mbt [hunk 4]
```diff
@@ -925,3 +925,264 @@ pub fn BigInt::to_octets(self : BigInt, length? : Int) -> Bytes {
...
```
Comment (peter-jerry-ye) [inline builtin/bigint_nonjs.mbt:1086]:
```suggestion
  }

  // return the result
```

File: builtin/bigint_nonjs.mbt [hunk 5]
```diff
@@ -925,3 +925,264 @@ pub fn BigInt::to_octets(self : BigInt, length? : Int) -> Bytes {
...
```
Comment (peter-jerry-ye) [inline builtin/bigint_nonjs.mbt:1173]:
```suggestion
  }

  // return the result
```

File: builtin/bigint_test.mbt [hunk 6]
```diff
@@ -557,3 +557,93 @@ test "to_hex" {
...
+  let a = BigInt::from_string("0")
+  let b = BigInt::from_string("-10")
+  inspect!(a ^ b, content="-10")
+}
+
+test "lxor_5" {
+  let a = BigInt::from_string("0")
+  let b = BigInt::from_hex("ffffffffffffffff")
+  inspect!(a ^ b, content="18446744073709551615")
+}
```
Comment (peter-jerry-ye) [inline builtin/bigint_test.mbt:649]:
```suggestion
}

```

**Review by peter-jerry-ye — CHANGES_REQUESTED** [review]
These should make the format test pass

**Review by peter-jerry-ye — APPROVED** [review]

**Comment by peter-jerry-ye** [comment]
Thank you for your contribution

**Pull request merged**